[FEAT] 아카이브 업로드 기능 구현

WhiteHyun opened the pull request:
## 📌 PR 요약

### 🌱 작업한 내용

- 업로드를 구현하기 위해 Network Service 일부 구현
   - ImageService의 deleteImage 구현

- UseCase 추가
   - UploadImageUseCase: 이미지 업로드
   - DeleteImageUseCase: 이미지 삭제
   - UploadArchiveUseCase: 아카이브 업로드
   - EditArchiveUseCase: 아카이브 수정

- 그 외 ArchiveUploadViewModel과 ArchiveUploadViewController간 파이프라인 연동
- 이미지 사진 제거 기능도 구현하였으나, 서버에서 이미지 제거 API에 문제가 생겨 스크린샷으로는 보여드리지 못하는 점 양해 바랍니다.


### 🌱 PR 포인트

#### 사진 추가 플로우

- `사진 추가 셀 선택` -> `ViewController에게 알려줌` -> `BottomSheet present` -> `사진 선택` -> `선택된 사진을 ViewModel에게 전달` -> `ViewModel이 사진 업로드 API 호출` -> `응답으로 받은 이미지 주소를 archiveContents에 저장` -> `snapshot update` -> `Cell Registration에서 이미지 주소를 가지고 configure 호출` -> `Cell View Update`

#### 사진 제거 플로우

- `셀의 삭제 버튼 인식` -> `셀은 갖고있는 이미지 주소를 ViewModel에게 전달` -> `ViewModel의 이미지 삭제 API 호출(UseCase 사용)` -> `ViewModel이 갖고있는 archiveContents 중에 이미지 주소와 겹치는 값 삭제` -> `snapshot update` -> `cell 제거`

#### 업로드(또는 수정)버튼 활성화 플로우

1. `HeaderView가 갖고있는 textField가 변경됨` -> `ViewModel에게 알림(delegate)`
2. `업로드된 사진을 갖는 archiveContents가 업데이트 될 때마다 Subject로 계속 값을 emit`
3. `1번`과 `2번`을 묶어 전부 비어있지 않으면 버튼 활성화가 될 수 있도록 처리(combineLatest 사용)



## 📸 스크린샷

|영상|영상 이후 적용된 아카이브|
|:--:|:-:|
|![아카이브 업로드 시나리오](https://user-images.githubusercontent.com/57972338/233888126-472dd49d-15e8-4a3d-bd6b-727742ccdaf3.gif)|![](https://user-images.githubusercontent.com/57972338/233888870-8fee07af-33f2-4d2d-a0d0-784a71dfcbaa.png)|

## 📮 관련 이슈
- Resolved: #299

WhiteHyun committed:
[FEAT] Cell binding with PhotoBottomSheet in ArchiveUpload

WhiteHyun committed:
[ADD] Add UploadImageUseCase

WhiteHyun committed:
[FEAT] Implement image upload binding and update of cell view in one flow

바텀시트로부터 이미지를 받고난 이후 ViewModel에게 전달,
전달받은 이미지를 Upload UseCase에게 요청 후 받은 이미지 링크를 가지고
Cell을 업데이트하는 식으로 하나의 파이프라인을 구성하였습니다.

WhiteHyun committed:
[FEAT] Delete image API

WhiteHyun committed:
[FEAT] Implement DeleteImageUseCase and inject to ArchiveUploadViewModel

WhiteHyun committed:
[FEAT] Delete Image API Binding in ArchiveUploadViewModel

이미지 셀의 X버튼이 눌렸을 때 ViewModel에게 정보를 알리고,
ViewModel은 해당 정보를 가지고 이미지 제거 API를 요청해서 이미지를 제거합니다.
그 이후 제거된 이미지를 제외한 뷰만 보이도록 Snapshot을 업데이트합니다.

WhiteHyun committed:
[FEAT] button enabled binding with archive title and image's count in ArchiveUploadVM

WhiteHyun committed:
[FEAT] Add UploadArchiveUseCase

WhiteHyun committed:
[FEAT] Add EditArchiveUseCase

WhiteHyun committed:
[FEAT] Inject uploadArchiveUseCase to ArchiveUploadViewModel

ArchiveUploadViewModel에게 UploadArchiveUseCase를 프로퍼티를 갖게끔 처리하였습니다.
추가로 ArchiveUploadViewModelWithEditFactory에서는 아카이브 수정에 관한 API를 연결해야 하기에
`EditArchiveUseCaseAdapter`를 별도로 구현하여 연결시켰습니다.

WhiteHyun committed:
[FEAT] Implement complete button tapped binding with Upload API

완료 또는 수정 버튼이 눌리면 Use Case가 호출되면서 API 요청을 진행하고,
성공 시 popViewController를 수행할 수 있도록 바인딩 로직을 구현했습니다.

WhiteHyun committed:
[ADD] Add comments

WhiteHyun committed:
Merge branch 'develop' into feat/299-Archive/Upload

WhiteHyun committed:
[ADD] Add comments

soobin-k commented on PLUB/Sources/Views/Home/Archive/ViewModel/ArchiveUploadViewModel.swift:
```diff
@@ -150,6 +275,36 @@ private extension ArchiveUploadViewModel {
+    let itemsToAdd = Set(archivesContents).subtracting(uploadedItems).map { Item.picture($0) }
+    snapshot.appendItems(itemsToAdd)
+    
+    snapshot.moveItem(addPictureItem, afterItem: snapshot.itemIdentifiers.last!)
```
강제 언래핑은 지양했으면 좋겟습니다!

WhiteHyun replied:
수정하겠습니다 😭

soobin-k commented on PLUB/Sources/Views/Home/Archive/ViewModel/ArchiveUploadViewModelFactory.swift:
```diff
@@ -32,7 +51,10 @@ final class ArchiveUploadViewModelWithUploadFactory: ArchiveUploadViewModelFacto
   
   static func make(plubbingID: Int, archiveID: Int = -1) -> ArchiveUploadViewModel {
```
초깃값 -1 대신 옵셔널로 지정하면 어떨까요??

WhiteHyun replied:
그게 더 좋을 거 같아요! 얼른 수정할게요 ☺️

soobin-k reviewed (approved):
고생많으셨습니다 승현님!!

WhiteHyun committed:
[CHORE] Replace force unwrapping to guard-let statement

WhiteHyun committed:
[CHORE] Update archiveID type Int to Int? in ArchiveUploadViewModelWithEditFactory

UploadFactory에서 archiveID를 쓰지않아 -1이라는 쓰레기 값을 넣었지만,
옵셔널 형태로 정의하여 쓰이지 않는 파라미터는 nil로 처리하는 것이 더 정당하다고 느꼈고 이를 수정했습니다.

Co-authored-by: soobin-k
